File: 2025/day01.py
# Advent of Code 2025 | https://adventofcode.com/2025/day/1
# Copyright 2025 by Giovanni Squillero
# SPDX-License-Identifier: 0BSD

from icecream import ic

INPUT_FILE_NAME = 'day01-test.txt'
# INPUT_FILE_NAME = 'day01-input.txt'

TOTAL_TICKS = 100


# Just roll...
# Notez bien: No need to reset the dial (-1 % 100 == 99)
def part_one(file_name: str) -> int:
    """Turn dial and check position at the end"""

    dial = 50
    password = 0
    with open(file_name) as file:
        for line in file:
            dial += int(line[1:]) if line[0] == 'R' else -int(line[1:])
            if dial % TOTAL_TICKS == 0:
                password += 1
    return password


# Just roll, one click at a time.
def part_two(file_name: str) -> int:
    """Turn dial and check position after each tick"""

    dial = 50
    password = 0
    with open(file_name) as file:
        for line in file:
            step = +1 if line[0] == 'R' else -1
            for _ in range(abs(int(line[1:]))):
                dial += step
                if dial % TOTAL_TICKS == 0:
                    password += 1
    return password


def main():
    password = part_one(INPUT_FILE_NAME)
    ic(password)
    password = part_two(INPUT_FILE_NAME)
    ic(password)


if __name__ == '__main__':
    main()

File: 2025/day01.go
// Advent of Code 2025 | https://adventofcode.com/2025/day/1
// Copyright 2025 by Giovanni Squillero
// SPDX-License-Identifier: 0BSD

package main

import (
	"bufio"
	"log"
	"os"
	"strconv"
)

const FileName string = "day01-test.txt"

// const FileName string = "day01-input.txt"

func main() {
	// Slurp file
	turns := ReadFile(FileName)

	// Part 1
	dial := 50
	password := 0
	for _, turn := range turns {
		dial += turn
		if dial%100 == 0 {
			password += 1
		}
	}
	log.Printf("Password (part 1): %v\n", password)

	// Part 2
	dial = 50
	password = 0
	for _, turn := range turns {
		var tick, count int
		if turn > 0 {
			tick = 1
			count = turn
		} else {
			tick = -1
			count = -turn
		}
		for t := 0; t < count; t += 1 {
			dial += tick
			if dial%100 == 0 {
				password += 1
			}
		}
	}
	log.Printf("Password (part 2): %v\n", password)
}

func ReadFile(fileName string) []int {
	file, err := os.Open(FileName)
	if err != nil {
		log.Panicf("%v", err)
	}
	defer file.Close()

	var movements []int
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		t, _ := strconv.Atoi(line[1:])
		if rune(line[0]) == 'R' {
			movements = append(movements, t)
		} else {
			movements = append(movements, -t)
		}
	}
	return movements
}
